cpp/src/arrow/csv/column_decoder.cc

```diff
@@ -0,0 +1,365 @@
+      : ColumnDecoder(std::move(task_group)),
+        pool_(pool),
+        col_index_(col_index),
+        num_chunks_(-1),
+        next_chunk_(0) {}
+
+  void Append(const std::shared_ptr<BlockParser>& parser) override {
+    Insert(static_cast<int64_t>(chunks_.size()), parser);
+  }
+
+  void SetEOF(int64_t num_blocks) override {
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    DCHECK_EQ(num_chunks_, -1) << "Cannot change EOF";
+    num_chunks_ = num_blocks;
+
+    // If further chunks have been requested in NextChunk(), arrange to return nullptr
+    for (int64_t i = num_chunks_; i < static_cast<int64_t>(chunks_.size()); ++i) {
+      auto* chunk = &chunks_[i];
+      if (chunk->is_valid()) {
+        DCHECK(!IsFutureFinished(chunk->state()));
+        chunk->MarkFinished(std::shared_ptr<Array>());
+      }
+    }
+  }
+
+  Result<std::shared_ptr<Array>> NextChunk() override {
+    std::unique_lock<std::mutex> lock(mutex_);
+
+    if (num_chunks_ > 0 && next_chunk_ >= num_chunks_) {
+      return nullptr;  // EOF
+    }
+    PrepareChunkUnlocked(next_chunk_);
+    auto chunk_index = next_chunk_++;
+    return WaitForChunkUnlocked(chunk_index);
+  }
+
+ protected:
+  // XXX useful?
+  virtual std::shared_ptr<DataType> type() const = 0;
+
+  Result<std::shared_ptr<Array>> WaitForChunkUnlocked(int64_t chunk_index) {
+    auto future = chunks_[chunk_index];  // Make copy because of resizes
+    mutex_.unlock();
+    auto maybe_array = future.result();
+    mutex_.lock();
+    return maybe_array;
+  }
+
+  void PrepareChunk(int64_t block_index) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    PrepareChunkUnlocked(block_index);
+  }
+
+  void PrepareChunkUnlocked(int64_t block_index) {
+    size_t chunk_index = static_cast<size_t>(block_index);
+    if (chunks_.size() <= chunk_index) {
+      chunks_.resize(chunk_index + 1);
+    }
+    if (!chunks_[block_index].is_valid()) {
+      chunks_[block_index] = Future<std::shared_ptr<Array>>::Make();
+    }
+  }
+
+  void SetChunk(int64_t chunk_index, Result<std::shared_ptr<Array>> maybe_array) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    SetChunkUnlocked(chunk_index, std::move(maybe_array));
+  }
+
+  void SetChunkUnlocked(int64_t chunk_index, Result<std::shared_ptr<Array>> maybe_array) {
+    auto* chunk = &chunks_[chunk_index];
+    DCHECK(chunk->is_valid());
+    DCHECK(!IsFutureFinished(chunk->state()));
+
+    if (maybe_array.ok()) {
+      chunk->MarkFinished(std::move(maybe_array));
+    } else {
+      chunk->MarkFinished(WrapConversionError(maybe_array.status()));
+    }
+  }
+
+  Status WrapConversionError(const Status& st) {
+    if (st.ok()) {
+      return st;
+    } else {
+      std::stringstream ss;
+      ss << "In CSV column #" << col_index_ << ": " << st.message();
+      return st.WithMessage(ss.str());
+    }
+  }
+
+  MemoryPool* pool_;
+  int32_t col_index_;
+
+  std::vector<Future<std::shared_ptr<Array>>> chunks_;
```

Comment:
Why does it need to preserve the ordering if each Chunk -> RecordBatch should be convertible in an isolated fashion (no mutex or shared state).

Reply:
I think users (at least in Python) will expect the CSV to be iterated in file order. That said we shouldn't keep a reference here once the chunk is fetched.

Reply:
This limitation exists due to how BaseStreamingReader does the iteration, i.e. ParseAndInsert push to the vector and DecodeNextChunk then picks it up later, both are called in the same method (`StreamingReader::ReadNext`) there must be a way to pass this information without using indirection and avoid any hidden shared state entirely.

```python
class BaseStreamingReader:
  def init(self, fd, schema=None):
     self.schema = set_or_infer_schema(fd, schema)
     self.chunker = chunker(...)

  def parse_tasks(self):
    """ A low-level interface that generates ParseTask. 

    The yielded task are ordered due to the nature of Chunker.
    There is no concurrency done except the thread-safety of Chunker, 
    e.g. protected by a mutex, the onus is always on the caller to synchronize
    the calls to this method if he needs them to be ordered, use a TaskGroup 
    to dispatch, and the logic re-ordering the futures' results.

    This is the interface that the Dataset FileFormat will use.
    """
    for buffer in self.chunker:
      # A closure (Buffer, ...) -> Result<RecordBatch>
      yield make_parse_task(buffer, decoders, self.schema)
  
class SerialStreamingReader(BaseStreamingReader):
  def record_batches(self):
    """
    High-level interface that generates RecordBatch.
    """
    for task in self.parse_tasks():
      yield task()
```

Reply:
One thing this is missing is that the type-inferring decoder has shared state by definition.